Pass f_locals as a dict rather than kwargs

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97703

Fixes https://github.com/pytorch/pytorch/issues/97688

This PR replaces accessing f_locals and f_globals directly, instead preferring to direct guards though dicts. L for locals, G for globals. This is reflected in our sources. 

This allows us to have what we call guard hygiene, where as stated in https://github.com/pytorch/pytorch/issues/97688 above, we do not reuse our polluted scope in CLOSURE_VARS. 

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

ezyang committed:
Pass f_locals as a dict rather than kwargs

Fixes https://github.com/pytorch/pytorch/issues/97688

One big problem is that instead of printing x < y we now print
`E["x"] < E["y"]` and now all of the tests wobbled and I'm mad.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

albanD commented on torch/_dynamo/guards.py:
```diff
@@ -87,17 +86,15 @@ def __init__(
+            scope = {'E': user_scope}
         else:
-            scope = dict()
+            scope = {'E': dict()}
```
nit: We should have a global define for the key.
Also any worry about conflicts? Why E ?

albanD reviewed (commented):
Code change sounds good.

voznesenskym committed:
Update on "Pass f_locals as a dict rather than kwargs"


Fixes https://github.com/pytorch/pytorch/issues/97688

One big problem is that instead of printing x < y we now print
`E["x"] < E["y"]` and now all of the tests wobbled and I'm mad.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Pass f_locals as a dict rather than kwargs"


Fixes https://github.com/pytorch/pytorch/issues/97688

One big problem is that instead of printing x < y we now print
`E["x"] < E["y"]` and now all of the tests wobbled and I'm mad.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented:
I split E into L and G to get around the fact that there are global imports we write to new names during `import_source` and we need to pipe globals down to check_fn. This is still easy to do, and the sources dictate which dict they should read from, so cross contamination is fine, as it is nested (top level collision is still protected via the guards `&` check). 

I have not yet fixed test wobble.

voznesenskym committed:
Update on "Pass f_locals as a dict rather than kwargs"


Fixes https://github.com/pytorch/pytorch/issues/97688

One big problem is that instead of printing x < y we now print
`E["x"] < E["y"]` and now all of the tests wobbled and I'm mad.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Pass f_locals as a dict rather than kwargs"


Fixes https://github.com/pytorch/pytorch/issues/97688

One big problem is that instead of printing x < y we now print
`E["x"] < E["y"]` and now all of the tests wobbled and I'm mad.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Pass f_locals as a dict rather than kwargs"


Fixes https://github.com/pytorch/pytorch/issues/97688

One big problem is that instead of printing x < y we now print
`E["x"] < E["y"]` and now all of the tests wobbled and I'm mad.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Pass f_locals as a dict rather than kwargs"


Fixes https://github.com/pytorch/pytorch/issues/97688

One big problem is that instead of printing x < y we now print
`E["x"] < E["y"]` and now all of the tests wobbled and I'm mad.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Pass f_locals as a dict rather than kwargs"


Fixes https://github.com/pytorch/pytorch/issues/97688

One big problem is that instead of printing x < y we now print
`E["x"] < E["y"]` and now all of the tests wobbled and I'm mad.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

jansel commented on torch/_dynamo/guards.py:
```diff
@@ -710,26 +711,31 @@ def compile_check_fn(
 
+        # TODO(voz): Make this a real util
+        def _print_true(*args):
+            print("PRINTING TRUE", args)
```
What is this?

jansel commented on torch/_dynamo/guards.py:
```diff
@@ -762,8 +768,9 @@ def id_ref(self, obj):
-    f_locals: Dict[str, object],
     index: int,
+    f_locals: Dict[str, object],
+    f_globals: Dict[str, object],
```
typing.Any?

jansel commented on torch/_dynamo/variables/import torch.py:
```diff
@@ -0,0 +1,11 @@
+    return x + inf
+
+
+print(f(torch.randn(2), 3))
```
What is this?

jansel commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -478,14 +476,16 @@ static PyObject* call_guard_fail_hook(
-      f_locals,
       PyLong_FromSize_t(index),
+      f_locals,
+      f_globals,
```
We shouldn't need to pass `f_globals` since it is known when we are creating the gaurds.  You can just bake it into a closure similar to what was done before.

Calling this function with anything other than the compile-time global scope wouldn't make sense.

ezyang replied:
I asked the same question to Voz and he responded in personal chat. I'll let Voz recap the arguments here though...

jansel commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -522,22 +522,24 @@ static PyObject* lookup(CacheEntry* e, THP_EVAL_API_FRAME_OBJECT *frame, CacheEn
-  } else {
-    valid = PyObject_Call(e->check_fn, noargs, f_locals);
+  if (f_locals == NULL) {
+      f_locals = frame->f_locals = PyDict_New();
```
When are we hitting this case?  Can you give an example and add a comment?

ezyang commented:
https://github.com/pytorch/pytorch/pull/97703#discussion_r1153408567 is the main comment

voznesenskym commented:
Thanks for review, not even close to done though, so going to ignore all the comments for now.

voznesenskym commented on torch/_dynamo/guards.py:
```diff
@@ -201,7 +199,7 @@ def ID_MATCH(self, guard: Guard):
     def NAME_MATCH(self, guard: Guard):
         obj = self.get(guard.name)
-        code = f"{self.arg_ref(guard)}.__name__ == {obj.__name__}"
+        code = f"{self.arg_ref(guard)}.__name__ == '{obj.__name__}'"
```
I don't know how this ever worked?

ezyang replied:
It appears NAME_MATCH is only called from one place:

```
        elif isinstance(value, HigherOrderOperator):
            return TorchHigherOrderOperator(
                value,
                guards=self.make_guards(
                    GuardBuilder.TYPE_MATCH, GuardBuilder.NAME_MATCH
                ),
            )
```

voznesenskym replied:
Yea, sure, I think I can do that in a separate PR to keep this cleaner. Let me do that real quick.

voznesenskym committed:
Update on "Pass f_locals as a dict rather than kwargs"


Fixes https://github.com/pytorch/pytorch/issues/97688

One big problem is that instead of printing x < y we now print
`E["x"] < E["y"]` and now all of the tests wobbled and I'm mad.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Pass f_locals as a dict rather than kwargs"


Fixes https://github.com/pytorch/pytorch/issues/97688

One big problem is that instead of printing x < y we now print
`E["x"] < E["y"]` and now all of the tests wobbled and I'm mad.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -522,17 +520,10 @@ static PyObject* lookup(CacheEntry* e, THP_EVAL_API_FRAME_OBJECT *frame, CacheEn
-  } else {
-    valid = PyObject_Call(e->check_fn, noargs, f_locals);
-  }
+  // TODO: In Python 3.9 can optimize this to PyObject_CallOneArg
```
```
// bpo-39245 made PyObject_CallOneArg() public (previously called
// _PyObject_CallOneArg) in Python 3.9.0a4
#if PY_VERSION_HEX < 0x030900A4 || defined(PYPY_VERSION)
PYCAPI_COMPAT_STATIC_INLINE(PyObject*)
PyObject_CallOneArg(PyObject *func, PyObject *arg)
{
    return PyObject_CallFunctionObjArgs(func, arg, NULL);
}
#endif
```
Nice

albanD replied:
Yes that capi compat is provided by python core devs and removes the need for almost all of these version check!

voznesenskym committed:
Update on "Pass f_locals as a dict rather than kwargs"


Fixes https://github.com/pytorch/pytorch/issues/97688

One big problem is that instead of printing x < y we now print
`E["x"] < E["y"]` and now all of the tests wobbled and I'm mad.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

jansel commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1626,7 +1626,7 @@ def evaluate_guards_for_args(self, placeholders, args):
         if guards:
             code = " and ".join(guards)
-            return eval(code, SYMPY_INTERP, dict(zip(arg_names, args)))
+            return eval(code, SYMPY_INTERP, {"G": dict(zip(arg_names, args))})
```
This doesn't seem right.  Should `G` point the global scope of the function so that if it is mutated the guards can see the changes?

voznesenskym replied:
This function, invoked through `eval_guards` is test only, as is `eval_guards` . `produce_guards` is the real production method it tests. We should refactor this into a test class, probably in another PR, or at least clarify that this is test only. 

I do not change the behavior of using the passed in {arg_names: args} mapping. As long as `GlobalSource` produces guards with `G` in them, this is more correct.

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -774,9 +773,14 @@ def enum_repr(value):
+    # TODO(voz): Do I need to check upstream guard_source against GuardSource.LOCAL(_NN_MODULE)
+    # or pass L/G?
+    name, val = enum_name.split(".")
+    local_name = f'L["{name}"].{val}'
```
This looks suspect

ezyang replied:
The general problem is there is no reason to believe that the enum is actually in scope (either neither locals nor globals). In other words, we need to apply the same playbook as in https://github.com/pytorch/pytorch/pull/96940

voznesenskym replied:
Hmm, I assumed this would be safe, but perhaps not. Alright, I can do something similar here.

ezyang commented on torch/_dynamo/source.py:
```diff
@@ -79,7 +79,7 @@ def reconstruct(self, codegen):
 
     def name(self):
-        return rename_implicit(f"random_value_{self.random_call_index}")
+        return f"random_value_{self.random_call_index}"
```
I guess this was also unhygienic before 🤔  Where does this even get bound

voznesenskym replied:
```
 if (
            self.is_supported_random()
            and all(k.is_python_constant() for k in args)
            and all(v.is_python_constant() for v in kwargs.values())
        ):
            args = [x.as_python_constant() for x in args]
            kwargs = {k: v.as_python_constant() for k, v in kwargs.items()}
            random_call_index = len(tx.random_calls)
            if random_call_index == 0:
                tx.output.initial_random_state = random.getstate()
            example_value = self.value(*args, **kwargs)
            source = RandomValueSource(random_call_index)
            tx.random_calls.append((self.value, args, kwargs))
            return VariableBuilder(tx, source).wrap_unspecialized_primitive(
                example_value
            )
```

Seems a little sus...

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -651,7 +655,6 @@ def source_ref(source):
     def compile_check_fn(
         self, local_builder, global_builder, guards_out, guard_fail_fn
     ):
-        assert not (set(local_builder.argnames) & set(global_builder.argnames))
```
How come?

voznesenskym commented:
> This broadly looks good, there's a bit of subtle L/G handling but since the tests are passing I guess it's right. Unfortunately, I can't approve myself. Do you want to reopen this PR under your name @voznesenskym for appropriate attribution too?

Sure thing

voznesenskym commented on torch/_dynamo/output_graph.py:
```diff
@@ -533,7 +535,7 @@ def compile_subgraph(
             val_to_names[stack_values[-1]] = list()
         for k, v in tx.symbolic_locals.items():
-            if isinstance(v.source, LocalSource) and v.source.name() == k:
+            if isinstance(v.source, LocalSource) and v.source.local_name == k:
```
We can't use name() because it references "G" or "L" while `k, v in tx.symbolic_locals.items()` knows it by its local_name

If you do not do this, you get spurious args in your graph. Let me know if you would like an example.

voznesenskym committed:
Update on "Pass f_locals as a dict rather than kwargs"


Fixes https://github.com/pytorch/pytorch/issues/97688

This PR replaces accessing f_locals and f_globals directly, instead preferring to direct guards though dicts. L for locals, G for globals. This is reflected in our sources. 

This allows us to have what we call guard hygiene, where as stated in https://github.com/pytorch/pytorch/issues/97688 above, we do not reuse our polluted scope in CLOSURE_VARS. 

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented:
https://github.com/pytorch/pytorch/pull/98107
